packages/ref-core/src/ref_core/constraints.py

```diff
@@ -0,0 +1,148 @@
+from typing import Protocol, runtime_checkable
+
+import pandas as pd
+from attrs import frozen
+from loguru import logger
+
+from ref_core.exceptions import ConstraintNotSatisfied
+
+
+@runtime_checkable
+class GroupValidator(Protocol):
+    """
+    A constraint that must be satisfied when executing a given metric run.
+
+    All constraints must be satisfied for a given group to be run.
+    """
+
+    def validate(self, group: pd.DataFrame) -> bool:
+        """
+        Validate if the constraint is satisfied by the dataset.
+
+        This is executed after the apply method to determine if the constraint is satisfied.
+        If the constraint is not satisfied, the group will not be executed.
+
+        Parameters
+        ----------
+        group
+            A group of datasets that is being validated.
+
+        Returns
+        -------
+        :
+            Whether the constraint is satisfied
+        """
+        ...
+
+
+@runtime_checkable
+class GroupOperation(Protocol):
+    """
+    An operation to perform on a group of datasets resulting in a new group of datasets.
+    """
+
+    def apply(self, group: pd.DataFrame, data_catalog: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perform an operation on the group of datasets.
+
+        If this operation modifies the group, the modified group should be returned.
+        If this operation fails, a ConstraintNotSatisfied exception should be raised.
+
+        Parameters
+        ----------
+        group
+            A group of datasets that is being validated.
```

Comment:
Can you specify somewhere (maybe just in words here or somewhere to not start down the path of pandas typing) what info the group df needs to contain, i.e. what columns are available, I assume it is long format?

Reply:
Yup sure. It seems like there needs to be some vocabulary definitions in the docs

FYI: A data catalog has a row per file, and the columns are defined by source type: https://github.com/CMIP-REF/cmip-ref/blob/9ea6e99ace037c2ef1a6ae26ce43d52915cd2961/packages/ref/src/ref/datasets/cmip6.py

Reply:
https://github.com/CMIP-REF/cmip-ref/blob/requirements/docs/how-to-guides/running-metrics-locally.py